doc/nom_recipes.md

```diff
@@ -0,0 +1,235 @@
+# Nom Recipes
+
+These are short recipes for accomplishing common tasks with nom.
+
+* [Whitespace](#whitespace)
+  + [Wrapper combinators that eat whitespace before and after a parser](#wrapper-combinators-that-eat-whitespace-before-and-after-a-parser)
+* [Comments](#comments)
+  + [`// C++/EOL-style comments`](#----c---eol-style-comments-)
+  + [`/* C-style comments */`](#----c-style-comments----)
+* [Identifiers](#identifiers)
+  + [`Rust-Style Identifiers`](#-rust-style-identifiers-)
+* [Literal Values](#literal-values)
+  + [Escaped Strings](#escaped-strings)
+  + [Integers](#integers)
+    - [Hexadecimal](#hexadecimal)
+    - [Octal](#octal)
+    - [Binary](#binary)
+    - [Decimal](#decimal)
+  + [Floating Point Numbers](#floating-point-numbers)
+
+## Whitespace
+
+
+
+### Wrapper combinators that eat whitespace before and after a parser
+
+```rust
+/// A combinator that takes a parser `inner` and produces a parser that also consumes both leading and 
+/// trailing whitespace, returning the output of `inner`.
+fn ws<'a, F: 'a, O, E: ParseError<&'a str>>(inner: F) -> impl Fn(&'a str) -> IResult<&'a str, O, E>
+  where
+  F: Fn(&'a str) -> IResult<&'a str, O, E>,
+{
+  move |i| {
+    delimited(
+      multispace0,
+      &inner,
+      multispace0
+    )(i)
+  }
```

Comment:
Why not just return `delimited(...)` directly?

Reply:
Vestigial code, I think. :)

Reply:
Ah, I remember now. I was emulating how Nom defines its combinators. To take a random example, look at [the implementation of char](https://github.com/Geal/nom/blob/master/src/character/complete.rs#L31).

As for why Nom uses a closure... I don't know.

Reply:
The reason Nom uses closures is that this is the only way to define primitive parsers and combinators. The Nom parser and combinator functions return "parser functions", which need custom logic, and that logic is encoded in the closure. Some of Nom's combinators could be implemented in terms of other combinators, but I believe this is avoided to improve performance. For example, the [`value()` combinator](https://github.com/Geal/nom/blob/1c34db743864455568ca9d30b2a90b26d94cca93/src/combinator/mod.rs#L535) could be written in terms of `map()`:
```rust
pub fn value<I, O1: Clone, O2, E: ParseError<I>, F>(
  val: O1,
  parser: F,
) -> impl FnMut(I) -> IResult<I, O1, E>
where
  F: Parser<I, O2, E>,
{
  map(parser, move |_| val.clone())
}
```

So it basically comes down to whether are implementing a new primitive combinator (which is rare, unless you are contributing to Nom), or using Nom's existing combinators to define a new combinator (which is usually the case, since a rich set of combinators are already provided).

Reply:
Here is how I would write this combinator (and I think it's a great example of a custom combinator):
```rust
fn ws<'a, O, E, F>(inner: F) -> impl FnMut(&'a str) -> IResult<&'a str, O, E>
where
  E: ParseError<&'a str>,
  F: FnMut(&'a str) -> IResult<&'a str, O, E>,
{
  delimited(multispace0, inner, multispace0)
}
```

Reply:
Thanks for the explanation!

Stylistically, having `E` in the `where` clause improves readability. I'll change that throughout. (See above for `FnMut` and `'a`.)